klippy/extras/ads1x1x.py

```diff
@@ -0,0 +1,358 @@
+
+        self._ppins = self._printer.lookup_object("pins")
+        self._ppins.register_chip(self.name, self)
+
+        self.pga = config.getchoice('pga', ADS1X1X_PGA, '4.096V')
+        self.mode = config.getchoice('mode', ADS1X1X_MODE, 'single')
+        # Comparators are not implemented, they would only be useful if the
+        # alert pin is used, which we haven't made configurable.
+        # But that wouldn't be useful for a normal temperature sensor anyway.
+        self.comp_mode = ADS1X1X_COMPARATOR_MODE['TRADITIONAL']
+        self.comp_polarity = ADS1X1X_COMPARATOR_POLARITY['ACTIVE_LO']
+        self.comp_latching = ADS1X1X_COMPARATOR_LATCHING['NON_LATCHING']
+        self.comp_queue = ADS1X1X_COMPARATOR_QUEUE['QUEUE_NONE']
+        self._i2c = bus.MCU_I2C_from_config(config, address)
+
+        self.mcu = self._i2c.get_mcu()
+
+        self._printer.add_object("ads1x1x " + self.name, self)
+        self._printer.register_event_handler("klippy:ready", \
+                                            self._handle_ready)
+        self._printer.register_event_handler("klippy:shutdown", \
+                                            self.reset_all_devices)
+
+        self._pins = {}
+        self._mutex = self._reactor.mutex()
+
+    def setup_pin(self, pin_type, pin_params):
+        if pin_type == 'adc':
+            pin = ADS1X1X_pin(self, pin_params)
+            if pin.pin in self._pins:
+                raise pins.error(
+                    '%s pin %s for chip %s is used multiple times' \
+                        % (self.chip, pin.pin, self.name))
+            self._pins[pin.pin] = pin
+            return pin
+        raise pins.error('Wrong pin or incompatible type: %s with type %s! ' % (
+            pin_params['pin'], pin_type))
+
+    def _handle_ready(self):
+        self.reset_all_devices()
+
+    def is_ready(self):
+        config = self._read_register(ADS1X1X_REG_POINTER['CONFIG'])
+        return bool((config & ADS1X1X_REG_CONFIG['OS_MASK']) == \
+                    ADS1X1X_OS['OS_IDLE'])
+
+    def sample(self, sensor):
+        with self._mutex:
+            pin_object = self._pins[sensor.pin]
+            sample = 0
+            try:
+                self._write_register(ADS1X1X_REG_POINTER['CONFIG'],
+                                     pin_object.config)
+                # The report time is 1 / sample_count * 4 to account for the 4
+                # possible inputs. So sample_count 16 on 1 input will result
+                # in 4 samples per second.
+                delay = 1 / self.samples_per_second_numeric
+                self._reactor.pause(self._reactor.monotonic() + delay)
+                while not self.is_ready():
+                    self._reactor.pause(self._reactor.monotonic() + 0.001)
+                sample = self._read_register(ADS1X1X_REG_POINTER['CONVERSION'])
+            except Exception:
+                logging.exception("ADS1X1X: error while sampling")
+                return None
```

Comment:
Can you provide a little more explanation for this error handling?  If I understand the code correctly, if there is some error in this interaction the code will just stop sending temperature updates to the higher-level code.  That sounds like that could cause a subtle failure (the higher level code will just assume the temperature hasn't changed for potentially extended periods).  I'm not sure what the best solution is, but perhaps invoking a shutdown should be considered.

Reply:
I could adjust it so that a failed sampling adds to the invalid_count and invokes a shutdown if it happens often enough in a row.

Reply:
Sounds good to me.